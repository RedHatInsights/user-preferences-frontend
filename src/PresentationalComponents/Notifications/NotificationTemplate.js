import React, { useCallback, useEffect, useState } from 'react';
import PropTypes from 'prop-types';
import { useFormApi } from '@data-driven-forms/react-form-renderer';
import WarningModal from '@patternfly/react-component-groups/dist/dynamic/WarningModal';
import useChrome from '@redhat-cloud-services/frontend-components/useChrome';
import FormButtons from '../shared/FormButtons';
import pathnames from '../../Utilities/pathnames';

const NotificationsTemplate = ({ schema, formFields }) => {
  const formApi = useFormApi();
  const [isVisibleDialog, setVisibleDialog] = useState(false);

  const { chromeHistory } = useChrome();
  const [triggerExit, setTriggerExit] = useState({
    confirmed: false,
    pathname: '',
    search: '',
  });

  const handleGoToIntendedPage = useCallback(
    (pathname, search) =>
      search.length > 0
        ? chromeHistory.push({ pathname, search })
        : chromeHistory.push(pathname),
    [chromeHistory]
  );

  useEffect(() => {
    const navigationAllowed =
      !formApi.getState().dirty ||
      triggerExit.confirmed ||
      triggerExit.search.length > 0;
    navigationAllowed &&
      handleGoToIntendedPage(triggerExit.pathname, triggerExit.search);

    const unblock = chromeHistory.block(({ location }) => {
      if (
        formApi.getState().dirty &&
        location.search?.length === 0 &&
        !(
          location.pathname?.includes(pathnames.notifications.link) &&
          location.pathname?.includes('user-preferences')
        )
      ) {
        setVisibleDialog(true);
      }
      setTriggerExit((obj) => ({
        ...obj,
        pathname: location.pathname,
        search: location.search,
      }));
      return navigationAllowed;
    });

    return () => {
      unblock();
    };
  }, [triggerExit.confirmed, triggerExit.pathname, triggerExit.search]);

  return (
    <>
      <WarningModal
        isOpen={isVisibleDialog}
        title="Unsaved changes"
        onConfirm={() => {
          setTriggerExit((obj) => ({
            ...obj,
            confirmed: true,
          }));
          setVisibleDialog(false);
        }}
        onClose={() => setVisibleDialog(false)}
      >
        Your page contains unsaved changes. Do you want to leave?
      </WarningModal>

      <form onSubmit={formApi.handleSubmit} style={{ display: 'contents' }}>
        {schema.title}
        {formFields}
        <FormButtons />
      </form>
    </>
  );
};

NotificationsTemplate.propTypes = {
  schema: PropTypes.shape({
    title: PropTypes.node,
  }),
  formFields: PropTypes.array,
};

export default NotificationsTemplate;
