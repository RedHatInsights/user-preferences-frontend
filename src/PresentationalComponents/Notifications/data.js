export const menuItems = {
  'application-services': {
    title: 'Application Services',
  },
  openshift: {
    title: 'OpenShift',
  },
  rhel: {
    title: 'Red Hat Enterprise Linux',
  },
  ansible: {
    title: 'Ansible',
  },
  //   edge: {
  //     title: 'Edge',
  //     application: 'notifications',
  //     resourceType: 'notification-preference?bundleName=edge',
  //     apps: [
  //       {
  //         label: 'Item1',
  //         schema: [],
  //       },
  //       {
  //         label: 'Item2',
  //         schema: [],
  //       },
  //       {
  //         label: 'Item3',
  //         schema: [],
  //       },
  //     ],
  //   },
  console: {
    title: 'Console',
    application: 'notifications',
  },
};
