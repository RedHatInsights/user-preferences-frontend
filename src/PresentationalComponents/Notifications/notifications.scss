.pref-notifications {
    &--container {
      height: 100%;
      background-color: var(--pf-global--BackgroundColor--100);
    }
    &--wrapper {
      display: block;
      height: 0;
    }
    &--grid {
      display: grid;
      grid-template: 
        "head head"
        "nav inputs" 
        "btn btn";
      grid-template-columns: 1fr 3fr;
      grid-template-rows: min-content auto 69px;
    }
    &--head {
      grid-area: head;
    }
    &--title {
      display: flex;      
      align-items: center;
      padding: var(--pf-global--spacer--lg) 0 var(--pf-global--spacer--sm) var(--pf-global--spacer--lg);
      .pf-v5-c-label { margin-left: var(--pf-global--spacer--md); }
    }
    &--subtitle {
      position: relative;
      border-bottom: 1px solid var(--pf-global--BorderColor--100);
      padding: 0  var(--pf-global--spacer--lg) var(--pf-global--spacer--lg) var(--pf-global--spacer--lg);
    }
    &--nav {
      grid-area: nav;
      border-right: 1px solid var(--pf-global--BorderColor--100);
    }
    &--inputs {
      grid-area: inputs;
      overflow-y: auto;
      padding: var(--pf-global--spacer--lg);
      background-color: var(--pf-global--BackgroundColor--100);
    }
  }

@media only screen and (max-width: 768px) {
  .pref-notifications {
    &--grid {
      grid-template: 
        "head"
        "nav" 
        "inputs"
        "btn";
      grid-template-rows: min-content auto auto 69px;
    }
  }
}
  