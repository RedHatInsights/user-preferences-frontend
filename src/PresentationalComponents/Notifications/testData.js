export const userPrefInitialState = {
  notificationsReducer: {
    bundles: {
      console: {
        name: 'notification-preferences',
        label: null,
        component: 'section',
        sections: [
          {
            name: 'sources',
            label: 'Sources',
            component: null,
            fields: [
              {
                name: null,
                label: null,
                component: null,
                fields: [
                  {
                    name: 'bundles[console].applications[sources].notifications[INSTANT]',
                    label: 'Instant notification',
                    description: 'Description.',
                    initialValue: true,
                    component: 'descriptiveCheckbox',
                    validate: [],
                    checkedWarning: 'Another warning',
                  },
                  {
                    name: 'is_subscribed',
                    label: 'email test',
                    description: 'Description.',
                    initialValue: true,
                    component: 'descriptiveCheckbox',
                    validate: [],
                    checkedWarning: 'Another warning',
                  },
                ],
              },
            ],
          },
        ],
      },
    },
  },
  emailReducer: {
    unsubscribe: {
      schema: [
        {
          fields: [
            {
              name: 'unsubscribe.from-all',
              isGlobal: true,
              label: 'Unsubscribe from all',
              initialValue: false,
              component: 'bulkSelectButton',
            },
          ],
        },
      ],
      loaded: true,
    },
    advisor: {
      schema: [
        {
          fields: [
            {
              name: 'is_subscribed',
              label: 'Weekly Report',
              title: 'Weekly report',
              description:
                "Subscribe to this account's Advisor Weekly Report email",
              helperText:
                "User-specific setting to subscribe a user to the account's weekly reports email",
              component: 'descriptiveCheckbox',
              isRequired: true,
              initialValue: false,
              isDisabled: false,
            },
          ],
        },
      ],
      loaded: true,
    },
  },
};

export const userPrefEmptyInitialState = {
  notificationsReducer: {
    bundles: {},
  },
  emailReducer: {},
};

export const calculateEmailConfigResponse = {
  advisor: {
    isVisible: {},
    url: '/user-preferences/',
    apiName: 'insights',
    bundle: 'rhel',
    title: 'Advisor',
  },
  unsubscribe: {
    localFile: 'data/unsubscribe.json',
    isVisible: true,
    title: '',
  },
};
