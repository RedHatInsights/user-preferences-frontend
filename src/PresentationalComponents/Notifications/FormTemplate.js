import React, { useEffect, useReducer, useRef, useState } from 'react';
import PropTypes from 'prop-types';

import FormSpy from '@data-driven-forms/react-form-renderer/form-spy';
import useFormApi from '@data-driven-forms/react-form-renderer/use-form-api';
import {
  ActionGroup,
  Bullseye,
  Button,
  Divider,
  EmptyState,
  EmptyStateBody,
  EmptyStateIcon,
  EmptyStateVariant,
  Flex,
  FlexItem,
  Form,
  Grid,
  GridItem,
  Menu,
  MenuContent,
  MenuGroup,
  MenuInput,
  MenuItem,
  MenuList,
  Spinner,
  TextInput,
  Title,
} from '@patternfly/react-core';
import { ButtonVariant } from '@patternfly/react-core/dist/js/components/Button/Button';
import { menuItems } from './data';
import './notifications.scss';
import { SearchIcon } from '@patternfly/react-icons';

/**
 * This id is requried to submit form by a button outside of the form element
 */
const MODAL_FORM_IDENTIFIER = 'modal-form';
export const CustomFormWrapper = (props) => (
  <Form {...props} id={MODAL_FORM_IDENTIFIER} />
);

const CustomButtons = ({ saveLabel, cancelLabel }) => {
  const { onCancel } = useFormApi();

  return (
    <FormSpy>
      {({ pristine, invalid, validating, submitting }) => (
        <div className="pf-c-form">
          <ActionGroup className="pf-u-mt-0">
            <Button
              ouiaId="primary-save-button"
              variant="primary"
              form={MODAL_FORM_IDENTIFIER}
              type="submit"
              isDisabled={pristine || validating || submitting || invalid}
            >
              {saveLabel}
            </Button>
            <Button
              ouiaId="secondary-cancel-button"
              variant="link"
              onClick={onCancel}
              id="cancel-modal"
            >
              {cancelLabel}
            </Button>
          </ActionGroup>
        </div>
      )}
    </FormSpy>
  );
};

CustomButtons.propTypes = {
  saveLabel: PropTypes.node,
  cancelLabel: PropTypes.node,
};

CustomButtons.defaultProps = {
  saveLabel: 'Save',
  cancelLabel: 'Cancel',
};

const renderPageHeading = (bundleTitle, sectionTitle) => (
  <React.Fragment>
    <Title
      headingLevel="h3"
      size="xl"
      className="pf-u-pb-xs pf-u-mt-lg"
      style={{ paddingLeft: 0 }}
    >
      {`${sectionTitle} | ${bundleTitle}`}
    </Title>
    Configure your {sectionTitle} notifications.
  </React.Fragment>
);

const reducer = (state, action) => {
  switch (action.type) {
    case 'changeTab':
      return {
        ...state,
        bundle: action.payload.bundle,
        section: action.payload.section,
      };
    default:
      throw new Error();
  }
};

const prepareSections = (bundles, searchValue = '') =>
  Object.entries(menuItems).reduce((acc, [bundleKey, value]) => {
    const sections = [
      ...(bundles[bundleKey]?.sections?.filter(
        (section) =>
          section.name
            .toLocaleLowerCase()
            .includes(searchValue.toLocaleLowerCase()) ||
          section.label
            .toLocaleLowerCase()
            .includes(searchValue.toLocaleLowerCase())
      ) || []),
    ];
    // Add email?
    return {
      ...acc,
      ...(sections?.length > 0
        ? {
            [bundleKey]: {
              ...value,
              sections,
            },
          }
        : {}),
    };
  }, {});

//   const calculateSection = (key, schema) => {
//     return getSection(key, schema, store?.[key], (isVisible) => {
//       const { ...config } = emailConfig;
//       if (isVisible === false) {
//         delete config[key];
//       } else {
//         config[key] = {
//           ...config[key],
//           isVisible,
//         };
//       }

//       setEmailConfig(config);
//     });
//   };

export const FormTemplate = ({ schema, formFields, isLoading, title }) => {
  const { handleSubmit } = useFormApi();
  const search = useRef(null);
  const [searchValue, setSearchValue] = useState();
  const [data, setData] = useState({});
  const mastheadHeight = useRef(0);
  useEffect(() => {
    mastheadHeight.current =
      window.innerHeight -
      document
        .getElementsByClassName('pf-c-page__main')[0]
        .getBoundingClientRect().height;
    const menuElement = document.getElementById('notifications-menu');

    resizeObserver.observe(menuElement);

    return () => {
      resizeObserver.unobserve(menuElement);
    };
  }, []);

  const resizeMenu = () => {
    const content = document.getElementById('notifications-menu-content');
    content.style.maxHeight =
      window.innerWidth > 768
        ? `${Math.min(
            document
              .getElementById('notifications-menu')
              .getBoundingClientRect().height -
              search.current?.getBoundingClientRect().height -
              1,
            window.innerHeight -
              mastheadHeight.current -
              title?.getBoundingClientRect().height -
              search.current?.getBoundingClientRect().height -
              1
          )}px`
        : '30vh';
  };

  const resizeObserver = new ResizeObserver(resizeMenu);

  const initialState = {
    bundle: Object.entries(schema)?.filter(
      (entry) => entry[1]?.sections.length > 0
    )?.[0]?.[0],
    section: Object.entries(schema)?.filter(
      (entry) => entry[1]?.sections.length > 0
    )?.[0]?.[1]?.sections[0].name,
  };
  const [state, change] = useReducer(reducer, initialState);

  useEffect(() => {
    setData(prepareSections(schema, searchValue));
  }, [schema, searchValue]);

  console.log(schema, formFields);

  return (
    // <form onSubmit={handleSubmit}>
    <Grid className="pf-u-h-100 pf-u-background-color-100">
      <GridItem
        id="notifications-menu"
        className="pf-m-3-col-on-md pref-notifications--menu"
      >
        <Menu isPlain isScrollable>
          <MenuInput ref={search} className="pf-u-mx-sm">
            <TextInput
              placeholder="Search applications"
              value={searchValue}
              aria-label="Filter menu items"
              iconVariant="search"
              type="search"
              onChange={(value) => setSearchValue(value)}
            />
          </MenuInput>
          <Divider />

          <MenuContent id="notifications-menu-content">
            {Object.keys(data)?.length > 0 || isLoading ? (
              Object.entries(data).map(([key, { title, sections = [] }]) => (
                <MenuGroup key={key} label={title} className="pf-u-px-sm">
                  <MenuList>
                    {sections.map((section) => (
                      <MenuItem
                        onClick={() =>
                          change({
                            type: 'changeTab',
                            payload: { bundle: key, section: section.name },
                          })
                        }
                        itemId={section.name}
                        key={section.name}
                        className="pf-u-px-sm"
                      >
                        {section.label}
                      </MenuItem>
                    ))}
                  </MenuList>
                </MenuGroup>
              ))
            ) : (
              <EmptyState
                variant={EmptyStateVariant.small}
                className="pf-u-mt-lg"
              >
                <EmptyStateIcon icon={SearchIcon} />
                <Title headingLevel="h4" size="lg">
                  No matching applications found
                </Title>
                <EmptyStateBody>
                  Adjust you filters and try again.
                </EmptyStateBody>
                <Button
                  variant={ButtonVariant.link}
                  onClick={() => setSearchValue('')}
                >
                  Clear filters
                </Button>
              </EmptyState>
            )}
          </MenuContent>
        </Menu>
      </GridItem>
      <GridItem className="pf-m-9-col-on-md pref-notifications--section">
        {!isLoading ? (
          <Flex className="pf-u-flex-direction-column pf-u-h-100">
            <FlexItem className="pf-u-mb-xl pf-u-ml-lg">
              {renderPageHeading(
                menuItems[state.bundle]?.title,
                schema[state.bundle]?.sections.find(
                  (section) => section.name === state.section
                )?.label
              )}
              {/** Do we want it to come together with the schema? */}
            </FlexItem>
            <FlexItem className="pf-u-flex-grow-1 pf-u-ml-lg">
              {schema.title}
              {formFields}
            </FlexItem>
            <div className="pref-notifications--buttons">
              <Button variant={ButtonVariant.primary} className="pf-u-mr-sm">
                Save
              </Button>
              <Button variant={ButtonVariant.link}>Cancel</Button>
            </div>
          </Flex>
        ) : (
          <Bullseye>
            <Spinner />
          </Bullseye>
        )}
      </GridItem>
    </Grid>
    // </form>
  );
};

export default FormTemplate;
