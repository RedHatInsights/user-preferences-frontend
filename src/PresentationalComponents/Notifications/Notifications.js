import React, { useEffect, useRef, useState } from 'react';
import omit from 'lodash/omit';
import useChrome from '@redhat-cloud-services/frontend-components/useChrome';
import { FormRenderer } from '@data-driven-forms/react-form-renderer';
import { componentMapper } from '@data-driven-forms/pf4-component-mapper';
import { Bullseye, Spinner, Text } from '@patternfly/react-core';
import { PageHeaderTitle } from '@redhat-cloud-services/frontend-components/PageHeader';
import { addNotification } from '@redhat-cloud-services/frontend-components-notifications/redux';
import { ScalprumComponent } from '@scalprum/react-core';
import { useDispatch, useSelector } from 'react-redux';
import { useStore } from 'react-redux';
import {
  getNotificationsSchema,
  saveNotificationValues,
} from '../../redux/actions/notifications-actions';
import { saveEmailValues } from '../../redux/actions/email-actions';
import { calculateEmailConfig } from '../../Utilities/functions';
import {
  BULK_SELECT_BUTTON,
  BulkSelectButton,
  DATA_LIST,
  DESCRIPTIVE_CHECKBOX,
  DataListLayout,
  DescriptiveCheckbox,
  FORM_TABS,
  INPUT_GROUP,
  InputGroup,
  LOADER,
  Loader,
  TAB_GROUP,
} from '../../SmartComponents/FormComponents';
import config from '../../config/config.json';
import FormTabs from './Tabs';
import FormTabGroup from './TabGroup';
import { prepareFields } from './utils';
import FormTemplate from './NotificationTemplate';
import './notifications.scss';

const Notifications = () => {
  const { auth } = useChrome();
  const dispatch = useDispatch();
  const titleRef = useRef(null);
  const [emailConfig, setEmailConfig] = useState({});
  const store = useStore();

  const emailPref = useSelector(({ emailReducer }) => emailReducer);
  const { bundles: notifPref, loaded } = useSelector(
    ({ notificationsReducer }) => ({
      ...notificationsReducer,
      bundles: Object.entries(notificationsReducer?.bundles || {})?.reduce(
        (acc, [key, value]) => ({
          ...acc,
          [key]: value,
        }),
        {}
      ),
    })
  );

  useEffect(() => {
    (async () => {
      await auth.getUser();
      setEmailConfig(calculateEmailConfig(config, dispatch));
      dispatch(getNotificationsSchema());
    })();
  }, []);

  const saveValues = (values, formApi) => {
    const notificationValues = {
      bundles: Object.entries(values.bundles).reduce(
        (acc, [bundleName, bundleData]) => ({
          ...acc,
          [bundleName]: {
            applications: Object.entries(bundleData.applications).reduce(
              (acc, [appName, appData]) => ({
                ...acc,
                [appName]: {
                  eventTypes: omit(appData.eventTypes, BULK_SELECT_BUTTON),
                },
              }),
              {}
            ),
          },
        }),
        {}
      ),
    };
    const promises = [dispatch(saveNotificationValues(notificationValues))];
    const submitEmail = formApi.getState().dirtyFields['is_subscribed'];
    // temporary submitting of RHEL Advisor email pref.
    if (submitEmail) {
      const { url, apiName } = emailConfig['advisor'];
      const action = saveEmailValues({
        application: 'advisor',
        values: { is_subscribed: values.is_subscribed },
        url,
        apiName,
      });
      promises.push(dispatch(action));
    }
    Promise.all(promises)
      .then(() => {
        submitEmail && setEmailConfig(calculateEmailConfig(config, dispatch));
        dispatch(getNotificationsSchema());
        dispatch(
          addNotification({
            dismissable: true,
            variant: 'success',
            title: 'Notification preferences successfully saved',
          })
        );
      })
      .catch(() => {
        dispatch(
          addNotification({
            dismissable: true,
            variant: 'danger',
            title: 'Notification preferences unsuccessfully saved',
          })
        );
      });
  };

  return loaded && Object.values(emailPref).every((value) => value.loaded) ? (
    <div id="notifications-container" className="pref-notifications--container">
      <div className="pref-notifications--wrapper">
        <div id="notifications-grid" className="pref-notifications--grid">
          <div ref={titleRef} className="pref-notifications--head">
            <PageHeaderTitle
              className="pref-notifications--title sticky"
              title="My Notifications"
            />
            <Text className="pref-notifications--subtitle">
              Opt in or out of receiving notifications, and choose how you want
              to be notified. Your Organization Administrator has configured
              which notifications you can or canâ€™t receive in their{' '}
              <a href={`/settings/notifications`}>Settings</a>.
              <ScalprumComponent
                module="./ConnectedTimeConfig"
                scope="notifications"
                store={store}
              />
            </Text>
          </div>

          <FormRenderer
            componentMapper={{
              ...componentMapper,
              [DESCRIPTIVE_CHECKBOX]: DescriptiveCheckbox,
              [BULK_SELECT_BUTTON]: BulkSelectButton,
              [LOADER]: Loader,
              [DATA_LIST]: DataListLayout,
              [INPUT_GROUP]: InputGroup,
              [FORM_TABS]: FormTabs,
              [TAB_GROUP]: FormTabGroup,
            }}
            FormTemplate={FormTemplate}
            schema={{
              fields: [
                {
                  component: FORM_TABS,
                  name: 'notifTabs',
                  titleRef,
                  bundles: notifPref,
                  fields: prepareFields(notifPref, emailPref, emailConfig),
                },
              ],
            }}
            onSubmit={saveValues}
          />
        </div>
      </div>
    </div>
  ) : (
    <Bullseye>
      <Spinner />
    </Bullseye>
  );
};

export default Notifications;
