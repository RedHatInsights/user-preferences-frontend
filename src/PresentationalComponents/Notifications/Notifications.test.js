/* eslint-disable react/prop-types */
import * as API from '../../api';
import * as EmailActions from '../../redux/actions/email-actions';
import * as NotificationsActions from '../../redux/actions/notifications-actions';
import * as functions from '../../Utilities/functions';

import {
  GET_NOTIFICATIONS_SCHEMA,
  SAVE_EMAIL_SCHEMA,
  SAVE_NOTIFICATION_SCHEMA,
} from '../../redux/action-types';
import {
  act,
  fireEvent,
  getAllByRole,
  getByRole,
  getByText,
  queryByText,
  render,
} from '@testing-library/react';
import { calculateEmailConfigResponse, userPrefInitialState } from './testData';

import { MemoryRouter } from 'react-router-dom';
import Notifications from './Notifications';
import { Provider } from 'react-redux';
import React from 'react';
import { ScalprumProvider } from '@scalprum/react-core';
import configureStore from 'redux-mock-store';
import promiseMiddleware from 'redux-promise-middleware';

const mockedNavigate = jest.fn();
const mockedLocation = jest.fn(() => ({}));

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useNavigate: () => mockedNavigate,
  useLocation: () => mockedLocation,
}));

jest.mock('@redhat-cloud-services/frontend-components/useChrome', () => {
  return () => ({
    chromeHistory: {
      push: jest.fn(),
      block: jest.fn(() => jest.fn()),
    },
    auth: { getUser: () => Promise.resolve() },
  });
});

const NotificationsWrapper = ({ store, children }) => (
  <ScalprumProvider
    api={{
      chrome: {
        getEnvironment: () => '',
        isProd: () => false,
        isBeta: () => false,
      },
    }}
  >
    <Provider store={store}>
      <MemoryRouter>{children}</MemoryRouter>
    </Provider>
  </ScalprumProvider>
);

describe('Notifications tests', () => {
  const middlewares = [promiseMiddleware];
  let mockStore;
  let initialState;

  const calculateEmailConfig = jest.spyOn(functions, 'calculateEmailConfig');
  const getApplicationSchema = jest.spyOn(API, 'getApplicationSchema');
  const getNotificationsSchemaSpy = jest.spyOn(
    NotificationsActions,
    'getNotificationsSchema'
  );
  const saveNotificationValues = jest.spyOn(
    NotificationsActions,
    'saveNotificationValues'
  );
  const saveEmailValues = jest.spyOn(EmailActions, 'saveEmailValues');
  const emptyResolve = Promise.resolve({
    fields: [],
  });

  beforeEach(() => {
    mockStore = configureStore(middlewares);
    initialState = userPrefInitialState;
  });

  afterEach(() => {
    calculateEmailConfig.mockReset();
    getApplicationSchema.mockReset();
    getNotificationsSchemaSpy.mockReset();
    saveNotificationValues.mockReset();
    saveEmailValues.mockReset();
    mockedNavigate.mockReset();
    mockedLocation.mockReset();
  });

  it('should render correctly', async () => {
    getApplicationSchema.mockImplementation(() => emptyResolve);
    getNotificationsSchemaSpy.mockImplementation(() => ({
      type: GET_NOTIFICATIONS_SCHEMA,
      payload: Promise.resolve({
        bundles: {},
      }),
    }));
    let wrapper;
    await act(async () => {
      wrapper = render(
        <NotificationsWrapper store={mockStore(initialState)}>
          <Notifications />
        </NotificationsWrapper>
      );
    });
    expect(wrapper.container).toMatchSnapshot();
  });

  it('should render empty state on filter', async () => {
    getApplicationSchema.mockImplementation(() => emptyResolve);
    getNotificationsSchemaSpy.mockImplementation(() => ({
      type: GET_NOTIFICATIONS_SCHEMA,
      payload: Promise.resolve({
        bundles: {},
      }),
    }));
    let wrapper;
    await act(async () => {
      wrapper = render(
        <Provider store={mockStore(initialState)}>
          <Notifications />
        </Provider>
      );
    });
    fireEvent.change(getByRole(wrapper.container, 'searchbox'), {
      target: { value: 'someText' },
    });
    expect(
      getByText(wrapper.container, 'No matching applications found')
    ).toBeVisible();
    fireEvent.click(getByText(wrapper.container, 'Clear filters'));
    expect(
      queryByText(wrapper.container, 'No matching applications found')
    ).toEqual(null);
    expect(getByText(wrapper.container, 'Console')).toBeVisible();
  });

  it('should submit correctly', async () => {
    getApplicationSchema.mockImplementation(() => emptyResolve);
    calculateEmailConfig.mockImplementation(() => calculateEmailConfigResponse);
    getNotificationsSchemaSpy.mockImplementation(() => ({
      type: GET_NOTIFICATIONS_SCHEMA,
      payload: Promise.resolve({
        bundles: {},
      }),
    }));
    saveNotificationValues.mockImplementation(() => ({
      type: SAVE_NOTIFICATION_SCHEMA,
      payload: Promise.resolve({}),
    }));
    saveEmailValues.mockImplementation(() => ({
      type: SAVE_EMAIL_SCHEMA,
      payload: Promise.resolve({}),
    }));
    let wrapper;
    await act(async () => {
      wrapper = render(
        <Provider store={mockStore(initialState)}>
          <Notifications />
        </Provider>
      );
    });
    expect(queryByText(wrapper.container, 'Save')).toEqual(null);
    fireEvent.click(getAllByRole(wrapper.container, 'checkbox').at(0));
    fireEvent.click(getAllByRole(wrapper.container, 'checkbox').at(1));
    expect(getByText(wrapper.container, 'Save')).toBeEnabled();
    fireEvent.click(getByText(wrapper.container, 'Cancel'));
    expect(queryByText(wrapper.container, 'Save')).toEqual(null);
    fireEvent.click(getAllByRole(wrapper.container, 'checkbox').at(0));
    fireEvent.click(getAllByRole(wrapper.container, 'checkbox').at(1));
    fireEvent.click(getByText(wrapper.container, 'Save'));
    expect(queryByText(wrapper.container, 'Save')).toEqual(null);
    expect(saveNotificationValues).toHaveBeenCalledTimes(1);
    expect(saveNotificationValues).toHaveBeenCalledWith({
      bundles: {
        console: {
          applications: {
            sources: {
              eventTypes: {},
            },
          },
        },
      },
    });
    expect(saveEmailValues).toHaveBeenCalledTimes(1);
    expect(saveEmailValues).toHaveBeenCalledWith({
      apiName: 'insights',
      application: 'advisor',
      url: '/user-preferences/',
      values: { is_subscribed: true },
    });
  });
});
