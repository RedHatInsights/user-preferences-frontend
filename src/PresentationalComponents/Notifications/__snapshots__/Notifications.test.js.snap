// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Notifications tests should render correctly 1`] = `
<div>
  <div
    class="pref-notifications--container"
    id="notifications-container"
  >
    <div
      class="pref-notifications--wrapper"
    >
      <div
        class="pref-notifications--grid"
        id="notifications-grid"
      >
        <div
          class="pref-notifications--head"
        >
          <h1
            class="pf-c-title pf-m-2xl pref-notifications--title sticky"
            data-ouia-component-id="OUIA-Generated-Title-1"
            data-ouia-component-type="PF4/Title"
            data-ouia-safe="true"
            widget-type="InsightsPageHeaderTitle"
          >
            My Notifications
          </h1>
          <p
            class="pref-notifications--subtitle"
            data-ouia-component-id="OUIA-Generated-Text-1"
            data-ouia-component-type="PF4/Text"
            data-ouia-safe="true"
            data-pf-content="true"
          >
            Opt in or out of receiving notifications, and choose how you want to be notified. Your Organization Administrator has configured which notifications you can or canâ€™t receive in their
             
            <a
              href="/settings/notifications"
            >
              Settings
            </a>
            .
            Scalprum Component Mock
          </p>
        </div>
        <form
          style="display: contents;"
        >
          <div
            class="pref-notifications--nav"
          >
            <div
              class="pf-c-menu pf-m-plain pf-m-scrollable"
              data-ouia-component-id="OUIA-Generated-Menu-1"
              data-ouia-component-type="PF4/Menu"
              data-ouia-safe="true"
            >
              <div
                class="pf-c-menu__search pf-u-mx-sm"
              >
                <input
                  aria-invalid="false"
                  aria-label="Filter menu items"
                  class="pf-c-form-control pf-m-search"
                  data-ouia-component-id="OUIA-Generated-TextInputBase-1"
                  data-ouia-component-type="PF4/TextInput"
                  data-ouia-safe="true"
                  placeholder="Search services"
                  type="search"
                  value=""
                />
              </div>
              <hr
                class="pf-c-divider"
              />
              <div
                class="pf-c-menu__content"
                id="notifications-menu-content"
              >
                <section
                  class="pf-c-menu__group pf-u-px-sm"
                >
                  <h1
                    class="pf-c-menu__group-title"
                    id=""
                  >
                    Console
                  </h1>
                  <ul
                    class="pf-c-menu__list"
                    role="menu"
                  >
                    <li
                      class="pf-c-menu__list-item"
                      role="none"
                    >
                      <button
                        class="pf-c-menu__item"
                        role="menuitem"
                        tabindex="0"
                        type="button"
                      >
                        <span
                          class="pf-c-menu__item-main"
                        >
                          <span
                            class="pf-c-menu__item-text"
                          >
                            Sources
                          </span>
                        </span>
                      </button>
                    </li>
                  </ul>
                </section>
              </div>
            </div>
          </div>
          <div
            class="pref-notifications--inputs"
          >
            <h3
              class="pf-c-title pf-m-xl pf-u-pb-xs"
              data-ouia-component-id="OUIA-Generated-Title-2"
              data-ouia-component-type="PF4/Title"
              data-ouia-safe="true"
            >
              Sources | Console
            </h3>
            <p
              class="pf-u-mb-md"
              data-ouia-component-id="OUIA-Generated-Text-2"
              data-ouia-component-type="PF4/Text"
              data-ouia-safe="true"
              data-pf-content="true"
            >
              Configure your 
              Sources
               notifications.
            </p>
            <div>
              <div
                class="pf-c-form"
              >
                <button
                  aria-disabled="false"
                  class="pf-c-button pf-m-secondary pref-c-bulk-select-button"
                  data-ouia-component-id="OUIA-Generated-Button-secondary-2"
                  data-ouia-component-type="PF4/Button"
                  data-ouia-safe="true"
                  id="bulk-select-sources"
                  name="bundles[console].applications[sources].eventTypes[BULK_SELECT_BUTTON]"
                  type="button"
                  value="true"
                >
                  Select
                   all
                </button>
                <div
                  class="pf-c-form__group pref-c-input-category"
                >
                  <div
                    class="pf-c-form__group-label"
                  >
                    <label
                      class="pf-c-form__label"
                    >
                      <span
                        class="pf-c-form__label-text"
                      >
                        Event notifications
                      </span>
                    </label>
                     
                  </div>
                  <div
                    class="pf-c-form__group-control"
                  >
                    <p
                      class="pf-u-pb-md"
                      data-ouia-component-id="OUIA-Generated-Text-4"
                      data-ouia-component-type="PF4/Text"
                      data-ouia-safe="true"
                      data-pf-content="true"
                    >
                      Select how would you like to receive notifications for each event.
                    </p>
                    <div
                      class="pf-c-form__group pref-c-input-group"
                    >
                      <div
                        class="pf-c-form__group-control"
                      >
                        <div
                          class="pf-c-check pref-c-descriptive-checkbox"
                        >
                          <input
                            aria-invalid="false"
                            class="pf-c-check__input"
                            data-ouia-component-id="OUIA-Generated-Checkbox-3"
                            data-ouia-component-type="PF4/Checkbox"
                            data-ouia-safe="true"
                            data-type="descriptive-checkbox"
                            id="descriptive-checkbox-bundles[console].applications[sources].notifications[INSTANT]"
                            name="bundles[console].applications[sources].notifications[INSTANT]"
                            type="checkbox"
                            value=""
                          />
                          <label
                            class="pf-c-check__label"
                            for="descriptive-checkbox-bundles[console].applications[sources].notifications[INSTANT]"
                          >
                            Instant notification
                          </label>
                          <span
                            class="pf-c-check__description"
                          >
                            <div />
                          </span>
                        </div>
                        <div
                          class="pf-c-check pref-c-descriptive-checkbox"
                        >
                          <input
                            aria-invalid="false"
                            class="pf-c-check__input"
                            data-ouia-component-id="OUIA-Generated-Checkbox-4"
                            data-ouia-component-type="PF4/Checkbox"
                            data-ouia-safe="true"
                            data-type="descriptive-checkbox"
                            id="descriptive-checkbox-is_subscribed"
                            name="is_subscribed"
                            type="checkbox"
                            value=""
                          />
                          <label
                            class="pf-c-check__label"
                            for="descriptive-checkbox-is_subscribed"
                          >
                            email test
                          </label>
                          <span
                            class="pf-c-check__description"
                          >
                            <div />
                          </span>
                        </div>
                        
                      </div>
                    </div>
                    
                  </div>
                </div>
              </div>
            </div>
          </div>
        </form>
      </div>
    </div>
  </div>
</div>
`;
